shared/src/main/scala/mlscript/codegen/typescript/TsTypegen.scala

```diff
@@ -39,18 +39,26 @@ final class TsTypegenCodeBuilder {
     *   scope for type parameter names
     */
   protected case class TypegenContext(
-      typeVarMapping: MutMap[TypeVar, SourceCode],
+      typeVarMapping: MutSortedMap[TypeVar, SourceCode],
       termScope: Scope,
       typeScope: Scope,
       isMethodDefintion: Boolean
   )
 
   object TypegenContext {
-    def apply(mlType: Type, isMethodDefintion: Boolean = false): TypegenContext = {
+
+    // define an ordering for type variables to store them as sorted
+    // use name hint if given otherwise use string representation of
+    // internal identifier to order
+    implicit val orderingTypeVar: Ordering[TypeVar] =
+      Ordering.by(tvar =>
+        tvar.nameHint.getOrElse(tvar.identifier.fold(_.toString, identity(_))))
```

Comment:
Can we move this to a `TypeVarImpl` trait like for `SimpleTermImpl`, so it can be reused in other places?

https://github.com/hkust-taco/mlscript/blob/795f31c2e6ec3555dfa1ea9065fe6bd59d0aa217/shared/src/main/scala/mlscript/helpers.scala#L329-L335

Besides, this ordering seems a bit broken: it will give the same order to two unequal things, as long as their `nameHint` is empty and they have `identifier.fold(_.toString, identity(_))` be the same thing.

Reply:
I'm assuming that within the same term/type defintion namehints will be unique. And I kept it here because this ordering is very specific typegen, other codegens/applications might want to hvae different ordering.

Should I still move this to TypeVar trait?

Reply:
What I meant to say is that I'm assuming the identifiers will be unique because they are adhoc type var names generated by type inference. Is it incorrect to assume this?

But besides this I'm not sure how else to combine namehints and identifiers into something unique.

I'll move the logic to TypeVarImpl. The only concern was that at a global level, some logic might try to combine type vars from different expressions in which case there might actually be a conflict. In case of typegen this won't happen because it's processing one expression at a time.

Reply:
I don't know what sort of conflict you're thinking about. An ordering is an ordering. If two things `compare` to `0`, it means they should be equal. That's not the case of your implementation. Again, think of the information you lose when you compute `identifier.fold(_.toString, identity(_))`.

> I'm assuming the identifiers will be unique because they are adhoc type var names generated by type inference. Is it incorrect to assume this?

This is irrelevant to the issue at hand AFAIK.

Reply:
Also, a hint: you can leverage the ordering of tuples if that makes it easier for you:

```scala
scala> import math.Ordered.orderingToOrdered

scala> (1, 2) compare (3, 4)
val res6: Int = -1
```

Reply:
Hmmm, I'm just realizing that sum types just don't have an implicit ordering, so the ordering must be defined by hand or by converting it to a product type, which is what you're suggesting.